File: comixed-services/src/main/java/org/comixedproject/service/comicpages/BlockedPageService.java
/*
 * ComiXed - A digital comic book library management application.
 * Copyright (C) 2021, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

package org.comixedproject.service.comicpages;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import lombok.extern.log4j.Log4j2;
import org.apache.commons.lang.time.DateFormatUtils;
import org.comixedproject.adaptors.csv.CsvAdaptor;
import org.comixedproject.messaging.PublishingException;
import org.comixedproject.messaging.comicpages.PublishBlockedPageRemovalAction;
import org.comixedproject.messaging.comicpages.PublishBlockedPageUpdateAction;
import org.comixedproject.messaging.library.PublishDuplicatePageListUpdateAction;
import org.comixedproject.model.comicpages.BlockedHash;
import org.comixedproject.model.net.DownloadDocument;
import org.comixedproject.repositories.comicpages.BlockedPageRepository;
import org.comixedproject.service.library.DuplicatePageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

/**
 * <code>BlockedPageService</code> applies business rules to instances of {@link BlockedHash}.
 *
 * @author Darryl L. Pierce
 */
@Component
@Log4j2
public class BlockedPageService {
  static final String PAGE_LABEL_HEADER = "Page Label";
  static final String PAGE_HASH_HEADER = "Hash Value";
  static final String PAGE_SNAPSHOT_HEADER = "Encoded Snapshot";

  @Autowired private BlockedPageRepository blockedPageRepository;
  @Autowired private CsvAdaptor csvAdaptor;
  @Autowired private PublishBlockedPageUpdateAction publishBlockedPageUpdateAction;
  @Autowired private PublishBlockedPageRemovalAction publishBlockedPageRemovalAction;
  @Autowired private PublishDuplicatePageListUpdateAction publishDuplicatePageListUpdateAction;
  @Autowired private DuplicatePageService duplicatePageService;

  /**
   * Returns all blocked pages.
   *
   * @return all blocked pages
   */
  public List<BlockedHash> getAll() {
    log.debug("Loading all blocked page records");
    return this.blockedPageRepository.getAll();
  }

  /**
   * Returns a single blocked page by its hash.
   *
   * @param hash the blocked page has
   * @return the blocked page
   * @throws BlockedPageException if no such record exists
   */
  public BlockedHash getByHash(final String hash) throws BlockedPageException {
    log.debug("Loading blocked page by hash: {}", hash);
    final BlockedHash result = this.blockedPageRepository.findByHash(hash);

    if (result == null) {
      throw new BlockedPageException("No such blocked page: hash=" + hash);
    }

    return result;
  }

  /**
   * Retrieves the set of hashes only.
   *
   * @return the blocked page hashes
   */
  public List<String> getHashes() {
    log.debug("Loading all blocked page hashes");
    return this.blockedPageRepository.getHashes();
  }

  private void doPublishDuplicatePageUpdates() {
    try {
      log.trace("Publishing duplicate page list");
      this.publishDuplicatePageListUpdateAction.publish(
          this.duplicatePageService.getDuplicatePages());
    } catch (PublishingException error) {
      log.error("Failed to publish duplicate page list", error);
    }
  }

  /**
   * Updates the details for a blocked page.
   *
   * @param hash the page hash
   * @param blockedHash the updated details
   * @return the updated record
   * @throws BlockedPageException if the hash is invalid
   */
  @Transactional
  public BlockedHash updateBlockedPage(final String hash, final BlockedHash blockedHash)
      throws BlockedPageException {
    final BlockedHash updatedPage = this.doBlockPageHash(hash, blockedHash);
    try {
      this.publishBlockedPageUpdateAction.publish(updatedPage);
    } catch (PublishingException error) {
      log.error("Failed to publish blocked page update", error);
    }
    this.doPublishDuplicatePageUpdates();
    return updatedPage;
  }

  public BlockedHash doBlockPageHash(final String hash, final BlockedHash source) {
    log.trace("Looking for existing blocked page record");
    BlockedHash pageRecord = this.blockedPageRepository.findByHash(hash);
    if (pageRecord == null) {
      log.trace("Creating new blocked page record");
      pageRecord = new BlockedHash("", hash, "");
    }
    if (source != null) {
      log.trace("Copying blocked page values");
      pageRecord.setLabel(source.getLabel());
    }
    log.trace("Saving updated blocked page");
    return this.blockedPageRepository.save(pageRecord);
  }

  /**
   * Adds a page type to the blocked page lis.
   *
   * @param hashes the page hashes
   */
  @Transactional
  public void blockPages(final List<String> hashes) {
    for (int index = 0; index < hashes.size(); index++) {
      final String hash = hashes.get(index);
      final BlockedHash blockedHashRecord = this.doBlockPageHash(hash, null);
      try {
        this.publishBlockedPageUpdateAction.publish(blockedHashRecord);
      } catch (PublishingException error) {
        log.error("Failed to publish blocked page update", error);
      }
    }
    this.doPublishDuplicatePageUpdates();
  }

  /**
   * Unblocked a set of page types by hash.
   *
   * @param hashes the hash list
   */
  @Transactional
  public void unblockPages(final List<String> hashes) {
    for (int index = 0; index < hashes.size(); index++) {
      final String hash = hashes.get(index);
      log.trace("Unblocking page hash: {}", hash);
      final BlockedHash blockedHashRecord = this.doUnblockPageHash(hash);
      if (blockedHashRecord != null) {
        log.trace("Publishing blocked page removal");
        try {
          this.publishBlockedPageRemovalAction.publish(blockedHashRecord);
        } catch (PublishingException error) {
          log.error("Failed to publish blocked page remove", error);
        }
      }
    }
    this.doPublishDuplicatePageUpdates();
  }

  public BlockedHash doUnblockPageHash(final String hash) {
    final BlockedHash entry = this.blockedPageRepository.findByHash(hash);
    if (entry == null) {
      log.trace("Page hash not blocked: {}", hash);
      return null;
    }
    log.trace("Deleting record");
    this.blockedPageRepository.delete(entry);
    return entry;
  }

  /**
   * Creates a CSV file containing the list of all blocked pages.
   *
   * @return the document
   */
  public DownloadDocument createFile() throws IOException {
    log.debug("Retrieving blocked pages");
    final List<BlockedHash> entries = this.blockedPageRepository.findAll();
    final byte[] content =
        this.csvAdaptor.encodeRecords(
            entries,
            (index, model) -> {
              if (index == 0) {
                return new String[] {PAGE_LABEL_HEADER, PAGE_HASH_HEADER, PAGE_SNAPSHOT_HEADER};
              } else {
                return new String[] {model.getLabel(), model.getHash(), model.getSnapshot()};
              }
            });
    return new DownloadDocument(
        String.format(
            "ComiXed Blocked Pages For %s.csv", DateFormatUtils.format(new Date(), "yyyy-MM-dd")),
        "text/csv",
        content);
  }

  /**
   * Processes a received file of blocked pages.
   *
   * @param inputStream the data stream
   * @return the updated list of blocked pages
   * @throws IOException if an error occurs with the data stream.
   */
  @Transactional
  public List<BlockedHash> uploadFile(final InputStream inputStream) throws IOException {
    this.csvAdaptor.decodeRecords(
        inputStream,
        new String[] {PAGE_LABEL_HEADER, PAGE_HASH_HEADER, PAGE_SNAPSHOT_HEADER},
        (index, row) -> {
          if (index > 0) {
            final String label = row.get(0);
            final String hash = row.get(1);
            final String snapshot = row.get(2);

            log.debug("Checking if blocked page already exists: hash={}", hash);
            var blockedPage = this.blockedPageRepository.findByHash(hash);
            if (blockedPage == null) {
              log.debug("Creating new blocked page record");
              this.doSaveRecord(label, hash, snapshot);
            }
          }
        });
    return this.blockedPageRepository.findAll();
  }

  @Transactional
  public void doSaveRecord(final String label, final String hash, final String snapshot) {
    final var blockedPage = new BlockedHash(label, hash, snapshot);
    this.blockedPageRepository.save(blockedPage);
  }

  /**
   * Deletes a set of entries based on their hashes.
   *
   * @param hashes the entry hashes
   * @return the deleted entries
   */
  @Transactional
  public List<String> deleteBlockedPages(final List<String> hashes) {
    log.debug("Deleting {} blocked page entr{}", hashes.size(), hashes.size() == 1 ? "y" : "ies");
    List<String> result = new ArrayList<>();
    hashes.forEach(
        hash -> {
          log.trace("Loading blocked page for hash: {}", hash);
          final BlockedHash entry = this.blockedPageRepository.findByHash(hash);
          if (entry != null) {
            log.trace("Deleting entry: id={}", entry.getId());
            this.blockedPageRepository.delete(entry);
            result.add(entry.getHash());
          }
        });
    log.trace("Returning list of deleted blocked pages");
    return result;
  }

  /**
   * Returns if the given has is blocked.
   *
   * @param hash the page hash
   * @return true if blocked
   */
  public boolean isHashBlocked(final String hash) {
    log.trace("Finding if hash is blocked: {}", hash);
    return this.blockedPageRepository.findByHash(hash) != null;
  }
}

File: comixed-webui/src/app/comic-pages/services/blocked-page.service.ts
/*
 * ComiXed - A digital comic book library management application.
 * Copyright (C) 2021, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Subscription } from 'webstomp-client';
import { Store } from '@ngrx/store';
import { LoggerService } from '@angular-ru/cdk/logger';
import { HttpClient } from '@angular/common/http';
import { selectMessagingState } from '@app/messaging/selectors/messaging.selectors';
import { WebSocketService } from '@app/messaging';
import { BlockedHash } from '@app/comic-pages/models/blocked-hash';
import {
  BLOCKED_HASH_LIST_REMOVAL_TOPIC,
  BLOCKED_HASH_LIST_UPDATE_TOPIC,
  DELETE_BLOCKED_PAGES_URL,
  DOWNLOAD_BLOCKED_PAGE_FILE_URL,
  LOAD_ALL_BLOCKED_PAGES_URL,
  LOAD_BLOCKED_PAGE_BY_HASH_URL,
  MARK_PAGES_WITH_HASH_URL,
  REMOVE_BLOCKED_STATE_URL,
  SAVE_BLOCKED_PAGE_URL,
  SET_BLOCKED_STATE_URL,
  UNMARK_PAGES_WITH_HASH_URL,
  UPLOAD_BLOCKED_PAGE_FILE_URL
} from '@app/comic-pages/comic-pages.constants';
import {
  blockedHashRemoved,
  blockedHashUpdated,
  loadBlockedHashList
} from '@app/comic-pages/actions/blocked-hash-list.actions';
import { interpolate } from '@app/core';
import { DeleteBlockedPagesRequest } from '@app/comic-pages/models/net/delete-blocked-pages-request';
import { SetBlockedStateRequest } from '@app/comic-pages/models/net/set-blocked-state-request';
import { MarkPagesWithHashRequest } from '@app/comic-pages/models/net/mark-pages-with-hash-request';
import { UnmarkPagesWithHashRequest } from '@app/comic-pages/models/net/unmark-pages-with-hash-request';

@Injectable({
  providedIn: 'root'
})
export class BlockedPageService {
  updateSubscription: Subscription;
  removalSubscription: Subscription;
  entriesLoaded = false;

  constructor(
    private logger: LoggerService,
    private http: HttpClient,
    private store: Store<any>,
    private webSocketService: WebSocketService
  ) {
    this.store.select(selectMessagingState).subscribe(state => {
      if (state.started && !this.updateSubscription) {
        this.logger.trace('Loading blocked page list');
        this.store.dispatch(loadBlockedHashList());
        this.logger.trace('Subscribing to blocked page updates');
        this.updateSubscription = this.webSocketService.subscribe<BlockedHash>(
          BLOCKED_HASH_LIST_UPDATE_TOPIC,
          entry => {
            this.logger.trace('Received blocked page update:', entry);
            this.store.dispatch(blockedHashUpdated({ entry }));
          }
        );
      }
      if (state.started && !this.removalSubscription) {
        this.logger.trace('Subscribing to blocked page removals');
        this.removalSubscription = this.webSocketService.subscribe<BlockedHash>(
          BLOCKED_HASH_LIST_REMOVAL_TOPIC,
          entry => {
            this.logger.trace('Received blocked page removal:', entry);
            this.store.dispatch(blockedHashRemoved({ entry }));
          }
        );
      }
      if (!state.started && !!this.updateSubscription) {
        this.logger.trace('Unsubscribing from blocked page updates');
        this.updateSubscription.unsubscribe();
        this.updateSubscription = null;
      }
      if (!state.started && !!this.removalSubscription) {
        this.logger.trace('Unsubscribing from blocked page removals');
        this.removalSubscription.unsubscribe();
        this.removalSubscription = null;
      }
    });
  }

  /**
   * Retrieves all blocked page entries.
   */
  loadAll(): Observable<any> {
    this.logger.trace('Loading all blocked pages');
    return this.http.get(interpolate(LOAD_ALL_BLOCKED_PAGES_URL));
  }

  /**
   * Loads a single blocked page by its hash.
   *
   * @param args.hash the page hash
   */
  loadByHash(args: { hash: string }): Observable<any> {
    this.logger.trace('Loading blocked page by hash:', args);
    return this.http.get(
      interpolate(LOAD_BLOCKED_PAGE_BY_HASH_URL, { hash: args.hash })
    );
  }

  /**
   * Saves a blocked page.
   *
   * @param args.entry the blocked page entry
   */
  save(args: { entry: BlockedHash }): Observable<any> {
    this.logger.trace('Saving blocked page:', args);
    return this.http.put(
      interpolate(SAVE_BLOCKED_PAGE_URL, { hash: args.entry.hash }),
      args.entry
    );
  }

  /**
   * Sets the blocked state for a page.
   * @param args.page the page
   * @param args.blocked the blocked state
   */
  setBlockedState(args: {
    hashes: string[];
    blocked: boolean;
  }): Observable<any> {
    if (args.blocked) {
      this.logger.trace('Blocking pages with hash:', args);
      return this.http.post(interpolate(SET_BLOCKED_STATE_URL), {
        hashes: args.hashes
      } as SetBlockedStateRequest);
    } else {
      this.logger.trace('Unblocking pages with hash:', args);
      return this.http.post(interpolate(REMOVE_BLOCKED_STATE_URL), {
        hashes: args.hashes
      } as SetBlockedStateRequest);
    }
  }

  /**
   * Downloads a file of blocked pages.
   */
  downloadFile(): Observable<any> {
    this.logger.trace('Download blocked pages file');
    return this.http.get(interpolate(DOWNLOAD_BLOCKED_PAGE_FILE_URL));
  }

  /**
   * Uploads a file of blocked pages.
   *
   * @param args.file the blocked page file
   */
  uploadFile(args: { file: File }): Observable<any> {
    this.logger.trace('Uploading blocked pages file:', args);
    const formData = new FormData();
    formData.append('file', args.file);
    return this.http.post(interpolate(UPLOAD_BLOCKED_PAGE_FILE_URL), formData);
  }

  /**
   * Deletes a set of blocked pages.
   *
   * @param args.entries the blocked page entries
   */
  deleteEntries(args: { entries: BlockedHash[] }): Observable<any> {
    this.logger.trace('Deleting blocked pages:', args);
    return this.http.post(interpolate(DELETE_BLOCKED_PAGES_URL), {
      hashes: args.entries.map(entry => entry.hash)
    } as DeleteBlockedPagesRequest);
  }

  markPagesWithHash(args: {
    hashes: string[];
    deleted: boolean;
  }): Observable<any> {
    if (args.deleted) {
      this.logger.trace('Marking pages with hashes:', args);
      return this.http.post(interpolate(MARK_PAGES_WITH_HASH_URL), {
        hashes: args.hashes
      } as MarkPagesWithHashRequest);
    } else {
      this.logger.trace('Unmarking pages with hashes:', args);
      return this.http.post(interpolate(UNMARK_PAGES_WITH_HASH_URL), {
        hashes: args.hashes
      } as UnmarkPagesWithHashRequest);
    }
  }
}
